### Day 33: Review Probability

#### 1. Một túi có 3 quả bóng đỏ và 5 quả bóng xanh. Xác suất rút được một quả bóng đỏ là bao nhiêu?

$P(\text{Đỏ})= \frac{\text{Số quả bóng đỏ}}{\text{Tổng số quả bóng}}=\frac{3}{3+5}=\frac{3}{8}=0.375$

In [2]:
# Số quả bóng màu đỏ
balls_red = 3
# Số quả bóng màu xanh
balls_blue = 5
# Tổng số quả bóng
total_ball = balls_red + balls_blue
# Xác suất để rút được một quả bóng màu đỏ
P_red = balls_red / total_ball
print("Xác suất rút được một quả bóng đỏ là: ", P_red)

Xác suất rút được một quả bóng đỏ là:  0.375


#### 2. Một con súc sắc sáu mặt được tung. Xác suất để số điểm là số chẵn
$P(\text{Số chẵn})= P(\text{Số 2})+P(\text{Số 4})+P(\text{Số 6})= \frac{1}{6}+\frac{1}{6}+\frac{1}{6}=\frac{3}{6}=0.5$


In [3]:
# số lượng mặt của một súc sắc
side = 6
# Xác suất của từng mặt số lẻ 
P_2 = P_4 = P_6 = 1 / side
# Xác suất để số điểm là số chẵn
P_even = P_2 + P_4 + P_6

print("Xác suất để số chẵn là:", P_even)

Xác suất để số chẵn là: 0.5


#### 3. Từ bộ bài 52 lá, xác suất để rút được một lá bài là quân bích

Ta có: 
- 4 Loại quân bài: Bích, Chuồn, Cơ, Rô(tổng có 4 loại )
- Với mỗi loại quân bài ta có số quân bài tương ứng: Át,2,3,4,5,6,7,8,9,10,J,Q,K (Tổng có 13 quân bài)
- Tổng công số lá ta được: 4(loại)* 13(quân)= 52 lá.


=> Vậy câu hỏi của đề bài ở đây có ý nghĩa là: Xác suất để rút một lá bài có số của quân bài từ Át,2,3,...,J,Q,K (Tổng có 13 quân bài) thuộc loại Bích

Vì số lượng số của quân từ Át,2,3,...,J,Q,K là như nhau đều bằng 1 ta được:

$P(X|C=bích) = \frac{\text{ Số quân là Át Bích}}{\text{Tổng số lá}}+\frac{\text{ Số quân là 2 bích}}{\text{Tổng số lá}}+...+\frac{\text{ Số quân là K bích }}{\text{Tổng số lá}} =\frac{\text{Tổng số Quân bích}}{\text{Tổng số lá}} =\frac{13}{52}=0.25$

Trong đó:
- X là số quân của lá bài
- C là loại bài  


In [4]:
type_card = 4
number_card = 13

total_card = type_card * number_card

P_X_C = number_card / total_card
print("Xác suất để rút đượcc một là bài là quân bích là:", P_X_C)

Xác suất để rút đượcc một là bài là quân bích là:  0.25


#### 4. Một hợp chứa 10 viên kẹo, trong đó có 7 viên kẹo vị dâu và 3 viên kẹo vị nho. Xác suất để lấy được viên kẹo vị nho?

$P(Nho)= \frac{\text{Số viên kẹo vị nho}}{\text{Tổng số viên kẹo}}= \frac{3}{10}=0.3$


In [6]:
keo_nho = 3
keo_dau = 7
tong_keo = keo_nho + keo_dau

P_nho = keo_nho / tong_keo

print("Xác suất để lấy đượ viên kẹo vị nho là:", P_nho)

Xác suất để lấy đượ viên kẹo vị nho là: 0.3


#### 5. Xác suất để một người chọn ngẫu nhiên sinh vào tháng 5
Vì mỗi tháng là độc lập và chỉ có một nên ta có

$P(\text{Tháng 5}) = \frac{1}{\text{Tổng số tháng}}=\frac{1}{12} \approx 0.08333$

In [7]:
number_month = 12
P_May = 1 / number_month

print("Xác suất để một người chọn ngẫu nhiên sinh vào tháng 5 là:", P_May)

Xác suất để một người chọn ngẫu nhiên sinh vào tháng 5 là: 0.08333333333333333


#### 6. Xác suất để một học sinh học giỏi toán (A) là 0.7. Xác suất để học sinh đó cũng học giỏi lý (B) là 0.6. Xác suất để học sinh giỏi cả hai môn (A và B) là 0.5. Tìm xác suất để học sinh học giỏi lý biết rằng học sinh đó giỏi toán.

$P(L|T) = \frac{P(T \cap L)}{P(T)}=\frac{0.5}{0.7}\approx 0.7143$

Trong đó L là Lý và T là Toán:
- $P(L|T)$ là Xác suất để học sinh học giỏi lý biết rằng học sinh đó giỏi toán
- $P(T \cap L)$ là xác suất để học sinh giỏi cả hai môn
- $P(T)$ là Xác suất để học sinh học giỏi toán

In [9]:
# Xác suất để học sinh giỏi cả hai môn
P_L_cap_T = 0.5

# xác suất để học sinh học giỏi toán
P_T = 0.7

# Xác suất để học sinh học giỏi lý biết rằng học sinh đó học giỏi toán
P_L_T = P_L_cap_T / P_T

print("Xác suất để học sinh giỏi lý biết rằng học sinh đó giỏi toán là:", round(P_L_T, 4))

Xác suất để học sinh giỏi lý biết rằng học sinh đó giỏi toán là: 0.7143


#### 7. Trong một lớp học, xác suất để học sinh đạt điểm A môn toán là 0.3, xác suất để đạt điểm A môn lý là 0.4. Biết rằng xác suất để dạt điểm A cả hai môn là 0.2, tìm xác suất để học sinh đạt điểm A môn toán biết rằng học sinh đó đã đạt điểm A môn lý

$P(\text{A_T|A_L})=\frac{P(\text{A_T \cap A_L})}{P(\text{A_L})}= \frac{0.2}{0.4}=0.5$

Trong đó :
- A_T là Điểm A môn Toán 
- A_L là Điểm A môn Lý
- $P(\text{A_T|A_L})$ là xác suất để học sinh đạt điểm A môn toán biết rằng học sinh đó đã đạt điểm A môn Lý
- $P(\text{A_T \cap A_L})$ là xác suất để đạt điểm A cả hai môn
- $P(\text{A_L})$ là xác suất để đạt điểm A môn lý 



In [10]:
P_AT = 0.3
P_AL = 0.4
P_AT_cap_AL = 0.2

P_AT_AL = P_AT_cap_AL / P_AL

print("Xác suất để học sính đạt điểm A môn toán biết rằng học sinh đó đã đạt điểm A môn lý là:", P_AT_AL)

Xác suất để học ính đạt điểm A môn toán biết rằng học sinh đó đã đạt điểm A môn lý là: 0.5


#### 8. Một túi có 5 quả táo và 3 quả cam. Rút ngẫu nhiên một quả và biết rằng quả đó là quả cam, xác suất để rút được một quả táo

Vì đã rút ra quả cam => Xác suất để rút ra được một quả táo bằng 0

#### 9. Một công ty có 60% nhân viên nam và 40% nhân viên nữ. Xác suất để một nhân viên nữ được thăng chức là 30%, và xác suất để một nhân viên nam được thăng chức là 20%. Tìm xác suất để một nhân viên được thăng chức biết rằng đó là nhân viên nữ.

Vì đã biết trước đó là nhân viên nữ =>
$P(\text{C|X="nữ"})= \text{30\%}$

#### 10. Xác suất để một sinh viên đậu một môn thi là 0.8. Xác suất để sinh viên đó đậu cả hai môn thi là 0.6. Tìm xác suất để sinh viên đậu môn thứ hai biết rằng đã đậu môn thứ nhất.

Gọi:
- A là sự kiện sinh viên đậu môn thứ nhất
- B là sự kiện sinh viên đậu môn thứ 2

$P(B|A) = \frac{P(A \cap B)}{P(A)}= \frac{0.6}{0.8}=0.75$


#### 11. Có một túi gồm 4 viên kẹo đỏ và 6 viên kẹo xanh. Một viên kẹo được chọn ngẫu nhiên và thấy là kẹo đỏ. Tìm xác suất túi còn lại có 5 viên kẹo đỏ và 5 viên kẹo xanh

Vì yêu cầu đề bài là tìm xác suất túi còn lại có 5 viên kẹo đỏ và 5 viên kẹo xanh nhưng trong túi chỉ có 4 viên kẹo đỏ và 6 viên kẹo xanh nên bài này không đủ hoặc thiếu dữ kiện để xác định

#### 12. Một bác sĩ cho biết có xác suất 5% để một bệnh nhân có bệnh khi kết quả xét nghiệm là dương tính. Xác suất để xét nghiệm dương tính khi bệnh nhân thực sự có bệnh là 95%, và xác suất để xết nghiệm dương tính khi bệnh nhân không có bệnh là 1%. Tìm xác suất thực sự có bệnh khi xét nghiệm dương tính.

A: Sự kiện bệnh nhân có bệnh
B: Sự kiện xét nghiệm dương tính
$\overline{A}$: Sự kiện bệnh nhân ko có bệnh
$\overline{B}$: Sự kiện xét nghiệm âm tính

Từ đề bài ta có:

$P(A) = 5\% $.
 
$P(B|A) = 95\%$.

$P(B|\overline{A})= 1\%$ .


Chúng ta cần tính P(B), xác suất xét nghiệm dương tính. Sử dụng định lý xác suất toàn phần:

$P(B)=P(B|A).P(A)+P(B|\overline{A}).P(\overline{A})$.

Trong đó:

$P(\overline{A})=1-P(A)=100\%-5\%=95\%$

Thay số vào

$P(B)= 0.95*0.05 + 0.01*0.95=0.057$

Bây giờ, áp dụng định lý Bayes:

$P(A|\overline{B})=\frac{P(B|A).P(A)}{P(B)}=\frac{0.95*0.05}{0.057} \approx 0.8333$
  

In [ ]:
P_A = 0.05
P_B_A = 0.95
P_B_A_ = 0.01
P_A_ = 1 - P_A

P_B = P_B_A * P_A + P_B_A_ * P_A_

P_A_B = (P_B_A * P_A) / P_B

print(f"Xác suát bệnh nhân thực sự có bệnh khi xét nghiệm dương tnh là khoảng {round(P_A_B, 4) * 100}%")

#### 13. Một người bạn có 60% khả năng xuất hiện tại một buổi tiệc nếu trời mưa và 30% nếu trời không mưa. Biết rằng xác suất trời mưa là 20%, xác suất người đó sẽ đến tiệc là bao nhiêu?

Gọi :
- $A$ là sự kiện người bạn xuất hiện tại buổi tiệc
- $\overline{A}$ là sự kiện người bạn ko xuất hiện tại buổi tiệc
- $B$ là sự kiện trời mưa
- $\overline{B}$ là sự kiện ko mưa

Từ đề bài ta có:

$P(A|B)=60\%$.

$P(A|\overline{B})=30\%$.

$P(B)=20\%$.

$P(\overline{B})=80\%$.

$P(A) = ?$.

Áp dụng định lý xác suất toàn phần

$P(A)=P(A|B).P(B) + P(A|\overline{B}).P(\overline{B})= 0.6*0.2+0.3*0.8= 0.36 $


In [5]:
P_A_B = 0.6
P_A_B_ = 0.3
P_B = 0.2
P_B_ = 0.8

P_A = P_A_B * P_B + P_A_B_ * P_B_

print(f"Vây với xác suất trời mưa là 20%, xác suất người đó sẽ đến tiệc là: {P_A * 100}%")

Vây với xác suất trời mưa là 20%, xác suất người đó sẽ đến tiệc là: 36.0%


#### 14. Trong một lớp học, 30% học sinh sử dụng máy tính xách tay, và trong số những học sinh này, 70% là nữ. Trong số những học sinh không sử dụng máy tính xách tay, 50% là nữ. Tìm xác suất để một học sinh sử dụng máy tính xách tay biết rằng học sinh đó là nữ:

Ta gọi: 
- $A$ là sự kiện học sinh là nữ
- $\overline{A}$ là sự kiện học sinh là nam
- $B$ là sự kiện học sinh sử dụng máy tính xác tay
- $\overline{B}$ là sự kiện học sinh không sử dụng máy tính xác tay

Theo đề bài ta có :

$P(B) =30\%$.

$P(\overline{B}) =100% - P(B) =70\%$.

$P(A|B) = 70\%$.

$P(A|\overline{B})=50%$

P(B|A) = ?

Áp dụng định lý xác suất toàn phần:

$P(A) = P(A|B).P(B) + P(A|\overline{B}).P(\overline{B})= 70\%*30\% + 70\%*50\% = 56\% = 0.56 $


Áp dụng định lý Bayes ta có: 

$P(B|A) = \frac{P(A|B)*P(B)}{P(A)} = \frac{70\%*30\%}{56\%}=37.5\% = 0.375 $


#### 15. Một bài kiểm tra ý tế có độ nhạy (sensitity) là 99%  và độ đặc hiệu (Specificity) là 95%. Xác suất mắc bệnh là 1%. Tìm xác suát để một người có kết quả dương tính thực sự mắc bệnh.

Ta gọi:
- $TP$: Số lượng mẫu cho kết quả dương tính là đúng
- $TN$: Số lượng mẫu cho kết quả dương tính là sai
- $FP$: Số lượng mẫu cho kết quả âm tính là đúng
- $FN$: SỐ lượng mẫu cho kết quả âm tính là sai
- $A$: Sự kiện một người mắc bệnh
- $\overline{A}$: Sự kiện một người không mắc bệnh
- $B$: Sự kiện một người có kết quả xét nghiệm là dương tính
- $\overline{B}$: Sự kiện một người có kết quả xét nghiệm là âm tính

Theo đề bài ta có:

(Độ nhạy ): Đây là xác suất một người có kết quả dương tính khi biết rằng người đó thực sự mắc bệnh

$Recall = P(B|A) =\frac{TP}{TP+FN}=99\%$

(Độ đặc hiệu): Đây là xác suát một người có kết quả âm tính khi biết rằng người đó thực sự không mắc bệnh

$Specificity =P(\overline{B}|\overline{A}) = \frac{TN}{TN+FP}=95\%$.

$P(B|\overline{A}) =100 -P(\overline{B}|\overline{A})=100\%-95\%=5\%$.

$P(A) =1\%$.

$P(\overline{A})= 100\%- P(A)= 99\%$.

P(A|B)=?


Áp dụng định lý xác suất toàn phần

$P(B)= P(B|A).P(A)+P(B|\overline{A}).P(\overline{A})= 99\% *1\% + 5\%*99\% =5.94\%=0.0594$.

Áp dụng định lý Bayes 

$P(A|B)=\frac{P(B|A).P(A)}{P(B)}=\frac{0.99*0.01}{0.0594} \approx 0.16667$


In [18]:
#Xác suất một người có kết quả dương tính khi biết rằng người đó thực sự mắc bệnh <=> P(B|A) = recall
recall = 0.99

#Xác suát một người có kết quả âm tính khi biết rằng người đó thực sự không mắc bệnh <=> P(B-|A-)= specificity
specificity = 0.95
P_B_A_ = 1 - specificity
P_A = 0.01
P_A_ = 1 - P_A
P_B = recall * P_A + P_B_A_ * P_A_

P_A_B = (recall * P_A) / P_B

print(f"Xác suất để một người có kết quả dương tính thực sự mắc bệnh là {round(P_A_B, 5) * 100.0}%")

Xác suất để một người có kết quả dương tính thực sự mắc bệnh là 16.667%


#### 16. cho biết phân phối chuẩn có trung bình là 0 và độ lệch chuẩn là 1. Tính xác suất để một biến ngẫu nhiên chuẩn nhận giá trị lớn hơn 1

Áp dụng công thức hàm phân phối tích lũy (CDF) để tính xác suất để một biến ngẫu nhiên chuẩn nhận giá trị nhỏ hơn hoặc bằng 1, ta có:

$F(x;\mu,\sigma) =\frac{1}{\sigma\sqrt{2\pi}}\int_{-\infty}^x exp(-\frac{ (t-\mu)^2}{2\sigma^2})dt$

<=> $F(1;0,1) =\frac{1}{\sqrt{2\pi}}\int_{-\infty}^1 exp(-\frac{ t^2}{2})dt$

<=> $F(1;0,1) =P(X<=1) \approx 0.84134 $

=> $P(X>1) =1 -P(X<=1) = 1-0.84134 = 0.15865 = 15.866 %$


In [21]:
from scipy.stats import norm

cdf_value = norm.cdf(1)

print(f"Giá trị của hàm CDF tại x=1 là: {cdf_value}")

print(f"Xác suất để một biến ngẫu nhiên chuẩn nhận giá trị lớn hơn 1 là {round(1 - cdf_value, 5) * 100}%")

Giá trị của hàm CDF tại x=1 là: 0.8413447460685429
Xác suất để một biến ngẫu nhiên chuẩn nhận giá trị lớn hơn 1 là 15.866%


#### 17. Một biến ngẫu nhiên X có phân phối chuẩn với trung bình 100 và độ lệch chuẩn 15. Tính xác suất để X nằm trong khoảng từ 85 đến 115

Đầu tiên là áp dụng công thức phân phối chuẩn hóa để quy mọi biến ngẫu nhiên phân phối chuẩn về dạng phân phối chuẩn tắc 

$Z= \frac{X-\mu}{\sigma}$

Với X=85:
$Z_1=\frac{85-100}{15}=-1$

Với X=115:
$Z_2=\frac{115-100}{15}=1$

Ta sử dụng bảng phân phối chuẩn tích lũy để tính P(-1<=Z<=1) thay thế cho P(85<=X<=115)

Theo bảng phân phối chuẩn tích lũy, xác suất để Z nằm giữa giá trị -1 và 1 là:

P(-1<=Z<=1) = P(Z<=1)-P(Z<=-1)

Từ bảng phân phối chuẩn tích lũy (hoặc hàm phân phối tích lũy chuẩn chuẩn hóa) ta có:

$P(Z<=1)\approx 0.8413$.

$P(Z<=-1)\approx 0.1587$. 


<=>P(-1<=Z<=1) = 0.8413-0.1587 = 0.6826 = P(85<=X<=115) 


In [29]:

cdf_value_1 = norm.cdf(85, loc = 100, scale = 15)
cdf_value_2 = norm.cdf(115, loc = 100, scale = 15)
probability = cdf_value_2 - cdf_value_1

print("Xác suất để X nằm trong khoảng từ 85-115 là:", round(probability * 100, 2))

Xác suất để X nằm trong khoảng từ 85-115 là: 68.27


#### 18. Xác suất để một biến ngẫu nhiên chuẩn nhận giá trị nhỏ hơn -2

Vì đầu bài ko đề cập đến mean và độ lệch chuẩn nên mặc định ta sẽ có mean =0 và độ lệch chuẩn = 1.


Từ bảng phân phối chuẩn tích lũy  ta có 

$F(-2;0,1) =\frac{1}{\sqrt{2\pi}}\int_{-\infty}^{-2} exp(-\frac{ t^2}{2})dt$

$F(-2;0,1) \approx 0.0228$

=>$P(X<-2)\approx 0.0228$ 

In [31]:
cdf_value_18 = norm.cdf(-2)

print("Xác suất để X <-2 là:", round(cdf_value_18 * 100, 2))

Xác suất để X <-2 là: 2.28


#### 19. Một bài kiểm tra có điểm trung bình là 75 và độ lệch chuẩn là 10. Tính xác suát để một học sinh đạt điểm trên 90 

Đầu tiên là áp dụng công thức phân phối chuẩn hóa để quy mọi biến ngẫu nhiên phân phối chuẩn về dạng phân phối chuẩn tắc (khi đó mean=0 và độ lệch chuẩn =1)

$Z= \frac{X- \mu}{\sigma}$

Với X = 90 : $Z =\frac{90-75}{10}=\frac{15}{10}=1.5 $

Ta sử dụng bảng phân phối chuẩn tích lũy để tính P(Z>=1.5) thay cho P(X>=90)

$P(Z>=1.5)=1-P(Z<1.5) =1- \frac{1}{\sqrt{2\pi}}\int_{-\infty}^{1.5} exp(-\frac{ t^2}{2})dt \approx 0.0668$

 

In [39]:
cdf_value_19 = norm.cdf(1.5)

print(f"Xác suất để một học sinh đạt điểm trên 90 là: {round((1 - cdf_value_19) * 100, 2)}%")

Xác suất để một học sinh đạt điểm trên 90 là: 6.68%


#### 20. Trong một kỳ thi, điểm của các sinh viên tuân theo phân phối chuẩn với trung binh 70 và độ lệch chuẩn 8. Tính xác suất để điểm số của một sinh viên nằm giữa 65 và 75.


Đầu tiên là ta áp dụng câu thức phân phối chuẩn hóa để quy mọi bến ngẫu nhiên phân phối chuẩn về dạng phân phối chuẩn tắc

$Z= \frac{X-\mu}{\sigma}$

Với X = 65:
$Z_1 = \frac{65-70}{8}=\frac{-5}{8} = -0.625 $

Với X = 75:
$Z_2 = \frac{75-70}{8}=\frac{5}{8} = 0.625 $

Ta sử dụng bảng phân phối chuẩn tích lũy để tính P(-0.625<=Z<=0.625) thay thế cho P(65<=X<=75)

Theo bảng phân phối chuẩn tích lũy, Xác suất để Z nằm giữa giá trị -0.625 và 0.625 là:

P(-0.625<=Z<=0.625) = P(Z<=0.625) - P(Z<=-0.625)

Từ bảng phân phối chuẩn tích lũy, ta có:

$P(Z<=0.625)\approx 0.7340$.

$P(Z<=-0.625)\approx 0.2660$. 

<=> P(-0.625<=Z<=0.625) = 0.7340 - 0.2660 = 0.468

In [38]:
cdf_value_20_1 = norm.cdf(65, loc = 70, scale = 8)
cdf_value_20_2 = norm.cdf(75, loc = 70, scale = 8)

probability_20 = cdf_value_20_2 - cdf_value_20_1

print(f"Xác suất để điểm số của một sinh viên nằm giữa 65 và 75 là: {round(probability_20 * 100, 2)}%")

Xác suất để điểm số của một sinh viên nằm giữa 65 và 75 là: 46.8%
